### PR TITLE
TimeRangePicker: Improve accessibility

### PR DESCRIPTION
**Pull request by tskarhed**
**What this PR does / why we need it**:
Improve keyboard navigation and screen reader support
- Add landmark markup
- Add input markup to the TimeRangeList
- Add keyboard focus style to TimeRangeList
- Improve TimePickerCalendar keyboard navigation
- Creates narrow screen accessible accordion

Does not include improvements to TimeZonePicker or the Calendar.

**Which issue(s) this PR fixes**:
Relates to #35613

### CODE REVIEW

**Comment by tskarhed** [comment]
It seems like a scrollbar appears in some place here. Not sure why.
Screenshot from 2021-07-22 11-41-23

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerContent.tsx [hunk 1]
```diff
@@ -218,19 +229,27 @@ const NarrowScreenForm: React.FC<FormProps> = (props) => {
...
         onClick={() => {
           if (!hideQuickRanges) {
             setCollapsedFlag(!collapsed);
           }
         }}
       >
-        <TimePickerTitle>Absolute time range</TimePickerTitle>
-        {!hideQuickRanges && <Icon name={!collapsed ? 'angle-up' : 'angle-down'} />}
-      </div>
+        <button
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerContent.tsx:241]: Needs an `aria-expanded` attribute

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx [hunk 2]
```diff
@@ -16,26 +18,52 @@ const getStyles = (theme: GrafanaTheme2) => {
...
   return (
-    <div className={cx(styles.container, selected && styles.selected)} onClick={() => onSelect(value)} tabIndex={-1}>
-      <span>{value.display}</span>
-    </div>
+    <li className={cx(styles.container, selected && styles.selected)}>
+      <input
+        className={styles.radio}
+        checked={selected}
+        name={name}
+        type="radio"
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx:59]: Radio groups causes automatic selection when navigating with keyboard - should we change to checkboxes or override the default behaviour?

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerContent.tsx [hunk 3]
```diff
@@ -218,21 +230,24 @@ const NarrowScreenForm: React.FC<FormProps> = (props) => {
...
-          if (!hideQuickRanges) {
-            setCollapsedFlag(!collapsed);
-          }
-        }}
-      >
-        <TimePickerTitle>Absolute time range</TimePickerTitle>
-        {!hideQuickRanges && <Icon name={!collapsed ? 'angle-up' : 'angle-down'} />}
-      </div>
+    <fieldset>
+      <h3 className={styles.header}>
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerContent.tsx:234]: Used an h3 here based on https://www.w3.org/TR/wai-aria-practices-1.1/examples/accordion/accordion.html

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.tsx [hunk 4]
```diff
@@ -89,11 +90,21 @@ export class UnthemedTimeRangePicker extends PureComponent<TimeRangePickerProps,
...
+            icon="angle-left"
+            narrow
+          />
+        )}
 
         <Tooltip content={<TimePickerTooltip timeRange={value} timeZone={timeZone} />} placement="bottom">
           <ToolbarButton
-            aria-label="TimePicker Open Button"
+            data-testid={selectors.components.TimePicker.openButton}
+            aria-label={`Time range editor ${formattedRange(value, timeZone)}`}
```
Comment (thisisobate) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.tsx:106]: This works great! 🎉👍

File: packages/grafana-e2e-selectors/src/selectors/components.ts [hunk 5]
```diff
@@ -6,11 +6,12 @@
 // prefix your selector string with 'data-test-id' so that when create the selectors we know to search for it on the right attribute
 export const Components = {
   TimePicker: {
-    openButton: 'TimePicker Open Button',
+    openButton: 'data-testidTimePicker Open Button',
```
Comment (thisisobate) [inline packages/grafana-e2e-selectors/src/selectors/components.ts:9]: `'data-testid TimePicker Open Button'`

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.tsx (hunk 4 above)
Comment (thisisobate) [reply packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.tsx:106]: I feel we need to use a lot more descriptive text in the Time range editor label text. Currently, the screen reader reads it as "Time range editor {any time range selected}" which might be a bit difficult for the user to comprehend. 

Rather, we could use something like `"Time range editor with current time range {any time range selected}"`

**Review by thisisobate — CHANGES_REQUESTED** [review]
Amazing work!  @tskarhed 🎉🎉

Left some thoughts for us to consider.

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerFooter.tsx [hunk 6]
```diff
@@ -59,13 +59,13 @@ export const TimePickerFooter: FC<Props> = (props) => {
...
           />
-        </div>
+        </section>
       </div>
     );
   }
 
   return (
-    <div className={style.container}>
+    <section aria-label="Time zone selection" className={style.container}>
```
[the remark this answers is not in the document]
Comment (tskarhed) [reply packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerFooter.tsx:68]: Tab navigation is usually done for components you can interact with. The panel header button should be enough for that. Adding section causes the screen reader to announce that you're entering or leaving a new region (with the aria label as its name). Using screen readers you can also do shortcuts between all the sections on the page.

The components inside the section are possible to navigate with a keyboard, so no reason why you would want to add another step of a non-interactable element.

**Review by thisisobate — APPROVED** [review]
LGTM! 🎉

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx (hunk 2 above)
Comment (thisisobate) [reply packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx]: I noticed some behavior with the checkboxes when interacting using the screen reader. The screen reader does not identify a `checked` box immediately it's been checked. It reads it as `'not checked'`. But then, clicking the TimePicker again shows the selected option as `checked`. 

I assume this is an issue with the screen reader software. Have you encountered anything like that? @tskarhed

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.test.tsx [hunk 7]
```diff
@@ -26,6 +26,6 @@ describe('TimePicker', () => {
       />
     );
 
-    expect(container.queryByLabelText(/timepicker open button/i)).toBeInTheDocument();
+    expect(container.queryByLabelText(/Time range editor/i)).toBeInTheDocument();
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.test.tsx:29]: Should we go with "Time range editor" or "Time range picker" here? Just want to make sure whatever it is, it's consistent.

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx [hunk 8]
```diff
@@ -16,26 +18,52 @@ const getStyles = (theme: GrafanaTheme2) => {
...
+   *  Input identifier. This should be the same for all options in a group.
+   */
+  name: string;
 }
 
-export const TimeRangeOption = memo<Props>(({ value, onSelect, selected = false }) => {
+export const TimeRangeOption = memo<Props>(({ value, onSelect, selected = false, name }) => {
   const styles = useStyles2(getStyles);
+  // In case there are more of the same timerange in the list
+  const id = value.display + Math.floor(Math.random() * 10);
```
Comment (kaydelaney) [inline packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimeRangeOption.tsx:51]: Could also use the `uuid` package

**Review by kaydelaney — APPROVED** [review]
Really awesome work here Tobias! I think this is a good start for the time range picker.
Some thoughts:
In Firefox it seems to take two tab presses to get from the time-picker button to highlighting the first relative time range (chrome seems to only need one).

It’s maybe a little surprising/confusing that the tab order starts with the relative time ranges - maybe starting with the “From” input would make more sense? Let me know what you think.

Also I’m not 100% sure about having all the relative time ranges be tab-navigable, just because having to go through all of them to get focus on the next toolbar item could be a little tedious, maybe the arrow keys could work better for those?
Lastly it might be nice to be able to press “Esc” and have the calendar close (not sure if role=“menu” semantics are applicable here, but something to look into).

Maybe you've read it before, but I've found https://w3c.github.io/aria-practices/#keyboard to be a big help.

In any case, awesome work. 👍

File: packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.test.tsx (hunk 7 above)
Comment (tskarhed) [reply packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker.test.tsx:29]: Good point, I'll change it to TimeRange picker instead

**Comment by tskarhed** [comment]
It's a bit frustrating with the way non-screen reader keyboard navigation handles radio buttons, which is the reason I changed to checkboxes. Then you're able to actually change the selection. The problem with radio buttons is also that the default and recommended behaviour is that the selection changes when moving with the arrow keys, causing the popup to close.

I changed so that the relative time ranges would come first in the order, but this is probably just a personal preference. The best thing we could do here is to ask a screen reader user.

**Pull request merged**